### PR TITLE
Add versions DAV endpoint

### LINKED ISSUE #4724: Public API to file versions
I would like to write a KDE 5 Dolphin integration to list and revert to previous versions of a file.
Is there an REST API available to access the version information like the sidebar does in the files app?

The documentation at https://docs.nextcloud.com/server/11/user_manual/files/version_control.html doesnt give any information on that.

### PR DESCRIPTION
**Pull request by rullzer**
Fixes #4724 

Basic functionality is there

### CODE REVIEW

File: apps/files_versions/lib/Sabre/VersionCollection.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+		return array_map(function (array $data) {
+			return new VersionFile($data, $this->userFolder->getParent());
+		}, $versions);
+	}
+
+	public function childExists($name): bool {
+
+
+		return false;
```
Comment (icewind1991) [inline apps/files_versions/lib/Sabre/VersionCollection.php]: Will this cause issues?

If not add a comment that it's safe

**Review by icewind1991 — DISMISSED** [review]

File: apps/files_versions/lib/Sabre/VersionHome.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	/** @var IRootFolder */
+	private $rootFolder;
+
+	public function __construct(array $principalInfo, IRootFolder $rootFolder) {
+		$this->principalInfo = $principalInfo;
+		$this->rootFolder = $rootFolder;
+	}
+
+	public function delete() {
+		throw new Forbidden('Permission denied to delete your trashbin');
```
[the remark this answers is not in the document]
Comment (rullzer) [reply apps/files_versions/lib/Sabre/VersionHome.php]: whoops

File: apps/files_versions/lib/Sabre/VersionCollection.php (hunk 1 above)
Comment (rullzer) [reply apps/files_versions/lib/Sabre/VersionCollection.php]: It should be fine I think.
But I just fixed it anyways. Not the most efficient approach but we can iterate later.

**Review by icewind1991 — APPROVED** [review]

**Comment by icewind1991** [comment]
For future work it might be worth looking into if we can expose available versions of a file as dav properties for a file in an efficient way

**Comment by rullzer** [comment]
@icewind1991 yes I agree. I think we can once we look into the version storage backend we talked about. But for now this is good enough I'd say ;) baby steps.

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by MorrisJobke** [comment]
Tests would also be nice ;)

**Comment by rullzer** [comment]
Yes yes. Will all come ;-)

**Comment by caco3** [comment]
Great work!
Looking forward to implement a KDE dolphin client for this.
Is there already an API documentation available?

**Pull request merged**